pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java

```diff
@@ -187,6 +189,51 @@ private void deleteFilesRecursivelyOnExit(Path path) {
                 }));
   }
 
+  /**
+   * Validates if the JVM memory is sufficient to run pipelines successfully. This validation helps
+   * the application to fail-fast rather than failing during batch execution.
+   */
+  private void validateHeapMemoryConfig(DataProperties dataProperties) {
+    long maxMemory = Runtime.getRuntime().maxMemory();
+    long minJVMMemoryRequired = minJVMMemory(dataProperties);
+    if (minJVMMemoryRequired > maxMemory) {
+      throw new IllegalConfigurationException(
+          String.format(
+              "Insufficient max JVM memory, required %s mb but provided %s mb",
+              memoryInMB(minJVMMemoryRequired), memoryInMB(maxMemory)));
+    }
+  }
+
+  /**
+   * The minimum JVM memory needed for the pipelines to run without fail. This is derived based on
+   * the number of parquet row groups that can be read or written in parallel by the pipeline
+   * threads. This is roughly derived by the below formula.
+   *
+   * <p>Minimum JVM Memory = Memory for JVM stack, perm files etc + (#Parallel Pipeline Threads *
+   * #Parallel Pipelines * (Parquet Row Group Size + buffer for parsing the parquet file))
+   */
+  private long minJVMMemory(DataProperties dataProperties) {
+    int rowGroupSize =
+        dataProperties.getRowGroupSizeForParquetFiles() > 0
+            ? dataProperties.getRowGroupSizeForParquetFiles()
+            : ParquetWriter.DEFAULT_BLOCK_SIZE;
+    int parallelism =
+        dataProperties.getNumThreads() > 0 ? dataProperties.getNumThreads() : DEFAULT_PARALLELISM;
+    // Temporary memory needed for decoding the read data from parquet files.
+    long tempBufferSize = (long) Math.max(40 * 1024 * 1024, 0.6 * rowGroupSize);
```

Comment:
Where do 40 and 0.6 come from? Can you please add that to the above comment? I am guessing that 1.6 is the inflation factor when Parquet content is loaded into memory, if that's the case, please consider replacing 0.6 with 1.6 here and remove `rowGroupSize` from line 226 below (it is  slightly more understandable).

Reply:
Simplified this to just have an inflation factor of 2 for the Parquet Row Groups. Documented this as well.